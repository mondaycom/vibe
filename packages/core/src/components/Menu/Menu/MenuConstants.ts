import { type ReactElement } from "react";

export type CloseMenuOption = {
  propagate?: boolean;
};

export type MenuChild = ReactElement & {
  type?: {
    isSelectable?: boolean;
    isMenuChild?: boolean;
    isMenu?: boolean;
    supportFocusOnMount?: boolean;
  };
};
