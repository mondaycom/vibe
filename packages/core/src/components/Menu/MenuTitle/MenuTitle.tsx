import cx from "classnames";
import React, { FC } from "react";
import { camelCase } from "lodash-es";
import { getStyle } from "../../../helpers/typesciptCssModulesHelper";
import { ComponentDefaultTestId, getTestId } from "../../../tests/test-ids-utils";
import Text from "../../Text/Text";
import { MenuTitleCaptionPosition as MenuTitleCaptionPositionEnum } from "./MenuTitleConstants";
import { MenuTitleCaptionPosition } from "./MenuTitle.type";
import { VibeComponentProps, withStaticProps } from "../../../types";
import styles from "./MenuTitle.module.scss";

export interface MenuTitleProps extends VibeComponentProps {
  /**
   * The caption text displayed alongside the title.
   */
  caption?: string;
  /**
   * The position of the caption relative to the title.
   */
  captionPosition?: MenuTitleCaptionPosition;
}

const MenuTitle: FC<MenuTitleProps> & {
  positions?: typeof MenuTitleCaptionPositionEnum;
  captionPositions?: typeof MenuTitleCaptionPositionEnum;
  isMenuChild?: boolean;
} = ({ className, caption = "", captionPosition = "bottom", id, "data-testid": dataTestId }: MenuTitleProps) => {
  const renderCaptionIfNeeded = () => {
    if (caption) {
      return (
        <label
          className={cx(styles.titleCaption, getStyle(styles, camelCase("title__caption--" + captionPosition)))}
          id={id}
          data-testid={dataTestId || getTestId(ComponentDefaultTestId.MENU_TITLE_CAPTION, id)}
        >
          {caption}
        </label>
      );
    }
  };
  return (
    <Text
      color="secondary"
      type="text2"
      className={cx(styles.title, className)}
      data-testid={dataTestId || getTestId(ComponentDefaultTestId.MENU_TITLE, id)}
    >
      {renderCaptionIfNeeded()}
    </Text>
  );
};

Object.assign(MenuTitle, {
  isMenuChild: true
});

export default withStaticProps(MenuTitle, {
  positions: MenuTitleCaptionPositionEnum,
  captionPositions: MenuTitleCaptionPositionEnum
});
