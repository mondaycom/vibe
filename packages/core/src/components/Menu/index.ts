export { default as Menu, type MenuProps } from "./Menu/Menu";
export { default as MenuItem, type MenuItemProps } from "./MenuItem/MenuItem";
export { default as MenuItemButton, type MenuItemButtonProps } from "./MenuItemButton/MenuItemButton";
export { default as MenuDivider, type MenuDividerProps } from "./MenuDivider/MenuDivider";
export { default as MenuTitle, type MenuTitleProps } from "./MenuTitle/MenuTitle";
export { default as MenuGridItem, type MenuGridItemProps } from "./MenuGridItem/MenuGridItem";

export * from "./MenuTitle/MenuTitle.type";
