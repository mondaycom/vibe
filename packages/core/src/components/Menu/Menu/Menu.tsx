import cx from "classnames";
import { SIZES } from "../../../constants/sizes";
import React, { forwardRef, ReactElement, useCallback, useEffect, useMemo, useRef, useState } from "react";
import useMergeRef from "../../../hooks/useMergeRef";
import useIsomorphicLayoutEffect from "../../../hooks/ssr/useIsomorphicLayoutEffect";
import useClickOutside from "../../../hooks/useClickOutside";
import useSubMenuIndex from "./hooks/useSubMenuIndex";
import useOnCloseMenu from "./hooks/useOnCloseMenu";
import useCloseMenuOnKeyEvent from "./hooks/useCloseMenuOnKeyEvent";
import useMenuKeyboardNavigation from "./hooks/useMenuKeyboardNavigation";
import useMouseLeave from "./hooks/useMouseLeave";
import { useAdjacentSelectableMenuIndex } from "./hooks/useAdjacentSelectableMenuIndex";
import { useFocusWithin } from "../../../hooks/useFocusWithin";
import usePrevious from "../../../hooks/usePrevious";
import { ElementContent, VibeComponent, VibeComponentProps, withStaticProps } from "../../../types";
import { CloseMenuOption, MenuChild } from "./MenuConstants";
import { getStyle } from "../../../helpers/typesciptCssModulesHelper";
import { getTestId } from "../../../tests/test-ids-utils";
import { ComponentDefaultTestId } from "../../../tests/constants";
import { useFocusOnMount } from "./hooks/useFocusOnMount";
import { useMenuId } from "./hooks/useMenuId";
import { generateMenuItemId } from "./utils/utils";
import styles from "./Menu.module.scss";

export interface MenuProps extends VibeComponentProps {
  /**
   * Size of the menu.
   */
  size?: (typeof SIZES)[keyof typeof SIZES];
  /**
   * The tab index of the menu.
   */
  tabIndex?: number;
  /**
   * ARIA label for accessibility.
   */
  ariaLabel?: string;
  /**
   * ARIA description ID.
   */
  ariaDescribedBy?: string;
  /**
   * If true, the menu will automatically focus on mount.
   */
  focusOnMount?: boolean;
  /**
   * Callback when a menu item gains focus.
   */
  onItemFocus?: (index: number) => void;
  /**
   * Controls the visibility of the menu.
   */
  isVisible?: boolean;
  /**
   * Callback triggered when the menu closes.
   */
  onClose?: (option: CloseMenuOption) => void;
  /**
   * Index of the focused menu item.
   */
  focusItemIndex?: number;
  /**
   * If true, this menu is a submenu.
   */
  isSubMenu?: boolean;
  /**
   * If true, event listeners will be attached to the document.
   */
  useDocumentEventListeners?: boolean;
  /**
   * Index of the item that should be focused when the menu mounts.
   */
  focusItemIndexOnMount?: number;
  /**
   * If true, enables scrolling within the menu.
   */
  shouldScrollMenu?: boolean;
  /**
   * The menu items.
   */
  children?: ElementContent;
}

const Menu: VibeComponent<MenuProps> & {
  isMenu?: boolean;
  supportFocusOnMount?: boolean;
  sizes?: typeof SIZES;
} = forwardRef(
  (
    {
      id,
      className,
      size = Menu.sizes.MEDIUM,
      tabIndex = 0,
      ariaLabel = "Menu",
      ariaDescribedBy,
      children: originalChildren,
      isVisible = true,
      onClose,
      onItemFocus,
      focusOnMount = false,
      focusItemIndex = -1,
      focusItemIndexOnMount = -1,
      isSubMenu = false,
      useDocumentEventListeners = false,
      shouldScrollMenu = false,
      "data-testid": dataTestId
    }: MenuProps,
    forwardedRef
  ) => {
    const ref = useRef(null);
    const mergedRef = useMergeRef(ref, forwardedRef);

    const overrideId = useMenuId(id);
    const splitMenuItemIconButtonRef = useRef<HTMLElement>(null);

    const [activeItemIndex, setActiveItemIndex] = useState(focusItemIndex);
    const [isInitialSelectedState, setIsInitialSelectedState] = useState(false);

    const children = useMemo(() => {
      const allChildren = React.Children.toArray(originalChildren) as MenuChild[];
      return allChildren.filter(child => {
        if (child.type.isMenuChild) return true;
        console.error(
          "Menu child must be a menuChild item (such as MenuItem, MenuDivider, MenuTitle, etc). This child is not supported: ",
          child
        );
        return false;
      });
    }, [originalChildren]);

    const updateActiveItemIndex = useCallback(
      (index: number) => {
        setActiveItemIndex(index);

        const activeChild = children[index];
        const ariaActiveDescendant = React.isValidElement(activeChild)
          ? activeChild?.props?.id || `${overrideId}-item-${index}`
          : undefined;
        if (ariaActiveDescendant) {
          ref?.current?.setAttribute("aria-activedescendant", ariaActiveDescendant);
        } else {
          ref?.current?.removeAttribute("aria-activedescendant");
        }
      },
      [children, overrideId]
    );

    const onSetActiveItemIndexCallback = useCallback(
      (index: number) => {
        onItemFocus && onItemFocus(index);
        updateActiveItemIndex(index);
        setIsInitialSelectedState(false);
      },
      [updateActiveItemIndex, onItemFocus]
    );

    const { setSubMenuIsOpenByIndex, hasOpenSubMenu, openSubMenuIndex, setOpenSubMenuIndex, resetOpenSubMenuIndex } =
      useSubMenuIndex();

    const onCloseMenu = useOnCloseMenu({
      setActiveItemIndex: onSetActiveItemIndexCallback,
      setOpenSubMenuIndex,
      onClose
    });

    useClickOutside({ ref, callback: () => onCloseMenu() });
    useCloseMenuOnKeyEvent({ hasOpenSubMenu, onCloseMenu, ref, onClose, isSubMenu, useDocumentEventListeners });

    const { getNextSelectableIndex, getPreviousSelectableIndex } = useAdjacentSelectableMenuIndex({
      children: children as ReactElement[]
    });
    useMenuKeyboardNavigation({
      hasOpenSubMenu,
      getNextSelectableIndex,
      getPreviousSelectableIndex,
      activeItemIndex,
      setActiveItemIndex: onSetActiveItemIndexCallback,
      isVisible,
      ref,
      useDocumentEventListeners
    });
    useMouseLeave({
      resetOpenSubMenuIndex,
      hasOpenSubMenu,
      ref,
      setActiveItemIndex: onSetActiveItemIndexCallback
    });
    useFocusOnMount({
      focusItemIndexOnMount,
      focusChildOnMount: children[focusItemIndexOnMount] as ReactElement,
      getNextSelectableIndex,
      updateActiveItemIndex,
      setIsInitialFocusSet: setIsInitialSelectedState
    });

    const onMouseMove = useCallback(() => {
      setIsInitialSelectedState(true);
    }, [setIsInitialSelectedState]);

    const previousHasOpenSubMenu = usePrevious(hasOpenSubMenu);
    useEffect(() => {
      if (hasOpenSubMenu || useDocumentEventListeners) return;
      if (activeItemIndex > -1 && previousHasOpenSubMenu) {
        // the submenu was just closed, so we want to focus the menu to capture keyboard events
        ref?.current?.focus();
      }
    }, [activeItemIndex, hasOpenSubMenu, previousHasOpenSubMenu, useDocumentEventListeners]);

    useIsomorphicLayoutEffect(() => {
      if (!focusOnMount || useDocumentEventListeners) return;
      requestAnimationFrame(() => {
        ref && ref.current && ref.current.focus();
      });
    }, [ref, focusOnMount, useDocumentEventListeners]);

    const { focusWithinProps } = useFocusWithin({
      onBlurWithin: () => {
        onCloseMenu && onCloseMenu();
      }
    });

    return (
      <ul
        onFocus={focusWithinProps?.onFocus}
        onBlur={focusWithinProps?.onBlur}
        id={overrideId}
        data-testid={dataTestId || getTestId(ComponentDefaultTestId.MENU, id)}
        className={cx(styles.menu, getStyle(styles, size), className)}
        ref={mergedRef}
        tabIndex={tabIndex}
        aria-label={ariaLabel}
        role="menu"
        aria-describedby={ariaDescribedBy}
        onMouseOver={onMouseMove}
      >
        {children &&
          React.Children.map(children, (child, index) => {
            return React.isValidElement(child)
              ? React.cloneElement(child, {
                  ...child?.props,
                  activeItemIndex,
                  index,
                  setActiveItemIndex: updateActiveItemIndex,
                  menuRef: ref,
                  resetOpenSubMenuIndex,
                  isParentMenuVisible: isVisible,
                  setSubMenuIsOpenByIndex,
                  hasOpenSubMenu: index === openSubMenuIndex,
                  closeMenu: onCloseMenu,
                  id: generateMenuItemId(overrideId, child, index),
                  useDocumentEventListeners,
                  isInitialSelectedState,
                  shouldScrollMenu,
                  getNextSelectableIndex,
                  getPreviousSelectableIndex,
                  isUnderSubMenu: isSubMenu,
                  splitMenuItemIconButtonRef
                })
              : null;
          })}
      </ul>
    );
  }
);

Object.assign(Menu, {
  isMenu: true,
  supportFocusOnMount: true
});

export default withStaticProps(Menu, {
  sizes: SIZES
});
