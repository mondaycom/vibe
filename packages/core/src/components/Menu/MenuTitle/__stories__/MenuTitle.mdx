import { Canvas, Meta } from "@storybook/blocks";
import * as MenuTitleStories from "./MenuTitle.stories";
import menuDo from "./Menu-do.svg";
import menuDont from "./Menu-dont.svg";

<Meta of={MenuTitleStories} />

# Menu Title

- [Overview](#overview)
- [Props](#props)
- [Variants](#variants)
- [Do’s and don’ts](#dos-and-donts)
- [Feedback](#feedback)

## Overview

<Canvas of={MenuTitleStories.Overview} />

### Import

```js
import { MenuTitle } from "@vibe/core";
```

## Props

<PropsTable />

## Variants

### Caption positions

<Canvas of={MenuTitleStories.CaptionPlacements} />

## Do’s and Don’ts

<ComponentRules
  rules={[
    {
      positive: {
        component: <img src={menuDo} alt="Menu title with an icon on the right" />,
        description: "Place the info icon inside an icon button and align it to the right."
      },
      negative: {
        component: <img src={menuDont} alt="Menu title with an icon on the left" />,
        description: "Don’t place the info icon to the left, top or bottom."
      }
    }
  ]}
/>
