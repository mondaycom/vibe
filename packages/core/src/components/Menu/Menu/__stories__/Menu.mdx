import { Meta } from "@storybook/blocks";
import {
  COMBOBOX,
  DROPDOWN,
  SPLIT_BUTTON
} from "../../../../storybook/components/related-components/component-description-map";
import {
  ComponentRuleDefaultWidth,
  ComponentRuleLargeWidth,
  ComponentRuleSimpleActions,
  ComponentRuleWithSearch,
  TipCombobox
} from "./Menu.stories.helpers";
import * as MenuStories from "./Menu.stories";

<Meta of={MenuStories} />

# Menu

- [Overview](#overview)
- [Props](#props)
- [Usage](#usage)
- [Variants](#variants)
- [Do’s and don’ts](#dos-and-donts)
- [Use cases and examples](#use-cases-and-examples)
- [Related components](#related-components)
- [Feedback](#feedback)

## Overview

A menu is a navigable contextual list of items that can be selected.

<Canvas of={MenuStories.Overview} />

### Import

```js
import { Menu } from "@vibe/core";
```

## Props

<PropsTable />

## Usage

<UsageGuidelines
  guidelines={[
    "A menu offers a list of actions or functions that a user can access.",
    "Menu height is dynamic according to the content it contains and its location on the screen.",
    "Closing menus can be done by selecting a value or clicking any where outside the menu.",
    "Menu items can include icons, radio buttons, and checkboxes.",
    "If a menu dropdown contains a mix of links and buttons, separate them with a content divider with links at the top and buttons at the bottom.",
    "Menu should contains at least two menu items."
  ]}
/>

<TipCombobox />

## Variants

### Sizes

<Canvas of={MenuStories.Sizes} />

### Menu with icons

<Canvas of={MenuStories.MenuWithIcons} />

## Do’s and Don’ts

<ComponentRules
  rules={[
    {
      positive: {
        component: <ComponentRuleSimpleActions />,
        description: "Use menus for simple actions."
      },
      negative: {
        component: <ComponentRuleWithSearch />,
        description: (
          <>
            Don’t place a search component near to menu for implement a filter capability. See{" "}
            <StorybookLink page="Components/Combobox">Combobox.</StorybookLink>
          </>
        )
      }
    },
    {
      positive: {
        component: <ComponentRuleDefaultWidth />,
        description: "Keep to the default menu width."
      },
      negative: {
        component: <ComponentRuleLargeWidth />,
        description: "Don’t change the width of the menu, only change the height."
      }
    }
  ]}
/>

## Use cases and examples

### Menu with sub menu

<Canvas of={MenuStories.MenuWithSubMenu} />

### Menu with 2-depth sub menu

<Canvas of={MenuStories.MenuWith2DepthSubMenu} />

### Menu with grid items and sub menu

Grid menu items are navigable with a keyboard as well

<Canvas of={MenuStories.MenuWithGridItemsAndSubMenu} />

## Related components

<RelatedComponents componentsNames={[COMBOBOX, DROPDOWN, SPLIT_BUTTON]} />
