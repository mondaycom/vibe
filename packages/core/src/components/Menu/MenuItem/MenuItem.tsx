import React, { AriaAttributes, ForwardedRef, ReactElement, forwardRef, useMemo, useRef } from "react";
import Tooltip, { TooltipProps } from "../../../components/Tooltip/Tooltip";
import Icon from "../../../components/Icon/Icon";
import useIsOverflowing from "../../../hooks/useIsOverflowing/useIsOverflowing";
import { SubIcon, VibeComponent, VibeComponentProps, withStaticProps } from "../../../types";
import { IconType } from "../../Icon";
import { CloseMenuOption, MenuChild } from "../Menu/MenuConstants";
import Label from "../../Label/Label";
import styles from "./MenuItem.module.scss";
import BaseMenuItem from "./components/BaseMenuItem/BaseMenuItem";
import MenuItemIcon from "./components/MenuItemIcon/MenuItemIcon";
import { TooltipPositions } from "../../Tooltip/Tooltip.types";
import { TooltipPositions as TooltipPositionsEnum } from "../../Tooltip/TooltipConstants";
import { SubmenuPosition } from "./MenuItem.types";

export interface MenuItemProps extends VibeComponentProps {
  /**
   * The title of the menu item.
   */
  title?: string;
  /**
   * The label displayed alongside the title.
   */
  label?: string | React.ReactElement<typeof Label>;
  /**
   * The icon displayed in the menu item.
   */
  icon?: SubIcon;
  /**
   * The type of icon.
   */
  iconType?: IconType;
  /**
   * The background color of the icon.
   */
  iconBackgroundColor?: string;
  /**
   * If true, the menu item is disabled.
   */
  disabled?: boolean;
  /**
   * The reason for disabling the item, shown in a tooltip.
   */
  disableReason?: string;
  /**
   * If true, the menu item is selected.
   */
  selected?: boolean;
  /**
   * Callback fired when the menu item is clicked.
   */
  onClick?: (event: React.MouseEvent | React.KeyboardEvent) => void;
  /**
   * The active item index in the menu.
   */
  activeItemIndex?: number;
  /**
   * Callback to set the active item index.
   */
  setActiveItemIndex?: (index: number) => void;
  /**
   * The index of the menu item.
   */
  index?: number;
  /**
   * The key of the menu item.
   */
  key?: string;
  /**
   * If true, the parent menu is visible.
   */
  isParentMenuVisible?: boolean;
  /**
   * Callback to reset the open submenu index.
   */
  resetOpenSubMenuIndex?: () => void;
  /**
   * If true, a submenu is open.
   */
  hasOpenSubMenu?: boolean;
  /**
   * Callback to open or close a submenu by index.
   */
  setSubMenuIsOpenByIndex?: (index: number, isOpen: boolean) => void;
  /**
   * If true, document event listeners are used for handling interactions.
   */
  useDocumentEventListeners?: boolean;
  /**
   * The tooltip content for the menu item.
   */
  tooltipContent?: string;
  /**
   * The position of the tooltip.
   */
  tooltipPosition?: TooltipPositions;
  /**
   * The delay in milliseconds before the tooltip shows.
   */
  tooltipShowDelay?: number;
  /**
   * Additional props for customizing the tooltip.
   */
  tooltipProps?: Partial<TooltipProps>;
  /**
   * Callback fired when the mouse leaves the item.
   */
  onMouseLeave?: (event: React.MouseEvent) => void;
  /**
   * Callback fired when the mouse enters the item.
   */
  onMouseEnter?: (event: React.MouseEvent) => void;
  /**
   * Class name applied to the icon wrapper.
   */
  iconWrapperClassName?: string;
  /**
   * If true, the menu item starts as selected.
   */
  isInitialSelectedState?: boolean;
  /**
   * If true, the menu scrolls to ensure visibility.
   */
  shouldScrollMenu?: boolean;
  /**
   * Function to close the menu with a given option.
   */
  closeMenu?: (option: CloseMenuOption) => void;
  /**
   * Reference to the menu container.
   */
  menuRef?: React.RefObject<HTMLElement>;
  /**
   * The submenu items, if applicable.
   */
  children?: MenuChild | MenuChild[];
  /**
   * If true, enables a split menu item interaction where the main area triggers an action,
   * while the icon button opens the submenu.
   */
  splitMenuItem?: boolean;
  /**
   * The label of the menu item for accessibility.
   */
  "aria-label"?: AriaAttributes["aria-label"];
  /**
   * The position of a submenu relative to the menu item.
   */
  submenuPosition?: SubmenuPosition;
  /**
   * If true, automatically repositions the submenu when its content changes.
   */
  autoAdjustOnSubMenuContentResize?: boolean;
}

export interface MenuItemTitleComponentProps extends Omit<MenuItemProps, "title"> {
  title: ReactElement;
  "aria-label": NonNullable<AriaAttributes["aria-label"]>;
}

const MenuItem: VibeComponent<MenuItemProps | MenuItemTitleComponentProps> & {
  iconType?: typeof Icon.type;
  isSelectable?: boolean;
  isMenuChild?: boolean;
  tooltipPositions?: typeof TooltipPositionsEnum;
} = forwardRef(
  (
    {
      className,
      iconWrapperClassName,
      title = "",
      label = "",
      icon = "",
      iconType,
      iconBackgroundColor,
      disabled = false,
      disableReason,
      selected = false,
      key,
      children,
      tooltipContent,
      tooltipPosition = "right",
      tooltipShowDelay = 300,
      tooltipProps,
      "aria-label": ariaLabel,
      ...baseMenuProps
    }: MenuItemProps | MenuItemTitleComponentProps,
    ref: ForwardedRef<HTMLElement>
  ) => {
    const titleRef = useRef();

    // if "title" is a component ariaLabel is mandatory
    const iconLabel = ariaLabel ?? (title as string);

    const isTitleHoveredAndOverflowing = useIsOverflowing({ ref: titleRef });
    const shouldShowTooltip = isTitleHoveredAndOverflowing || disabled || tooltipContent;

    const finalTooltipContent = useMemo(() => {
      if (disabled) return disableReason;
      if (tooltipContent) return tooltipContent;
      return title;
    }, [disableReason, disabled, title, tooltipContent]);

    const renderLabel = useMemo(() => {
      if (!label) return;
      if (typeof label === "string") {
        return <Label kind="line" text={label} />;
      }
      if (React.isValidElement(label) && label.type === Label) {
        return label;
      }
    }, [label]);

    return (
      <Tooltip
        content={shouldShowTooltip ? finalTooltipContent : null}
        position={tooltipPosition}
        showDelay={tooltipShowDelay}
        {...tooltipProps}
      >
        <BaseMenuItem
          key={key}
          ref={ref}
          subMenu={children}
          className={className}
          disabled={disabled}
          selected={selected}
          {...baseMenuProps}
        >
          {Boolean(icon) && (
            <MenuItemIcon
              icon={icon}
              type={iconType}
              label={iconLabel}
              disabled={disabled}
              selected={selected}
              backgroundColor={iconBackgroundColor}
              wrapperClassName={iconWrapperClassName}
            />
          )}
          <div ref={titleRef} className={styles.title}>
            {title}
          </div>
          {renderLabel}
        </BaseMenuItem>
      </Tooltip>
    );
  }
);

Object.assign(MenuItem, {
  isSelectable: true,
  isMenuChild: true
});

export default withStaticProps(MenuItem, {
  iconType: Icon.type,
  tooltipPositions: TooltipPositionsEnum
});
