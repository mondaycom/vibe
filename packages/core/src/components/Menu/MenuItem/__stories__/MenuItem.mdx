import { Meta } from "@storybook/blocks";
import { ComponentRuleWithLabelDo, ComponentRuleWithLabelDont } from "./MenuItem.stories.helpers";
import * as MenuItemStories from "./MenuItem.stories";

<Meta of={MenuItemStories} />

# Menu Item

Use menu item for drawing one options that displayed inside a menu.

<Canvas of={MenuItemStories.Overview} />

### Import

```js
import { MenuItem } from "@vibe/core";
```

## Props

<PropsTable />

## Accessibility

<UsageGuidelines
  guidelines={[
    <>
      Use the <code>ariaLabel</code> prop when you need to provide a custom accessible name that differs from the
      visible title text, especially for icon-only menu items.
    </>,
    <>
      For menu items with submenus, ensure the submenu content is properly structured with descriptive menu item titles
      for clear navigation hierarchy.
    </>,
    <>
      Use the <code>disabled</code> prop appropriately with a <code>disableReason</code> to provide context for why a
      menu item cannot be selected.
    </>
  ]}
/>

## Variants

### States

<Canvas of={MenuItemStories.States} />

### Icons

<Canvas of={MenuItemStories.Icons} />

### Label

<Canvas of={MenuItemStories.Label} />

### SubMenu

<Canvas of={MenuItemStories.SubMenu} />

### Overflow

<Canvas of={MenuItemStories.Overflow} />

### Tooltips

<Canvas of={MenuItemStories.TooltipStory} />

## Do's and Donâ€™ts

<ComponentRules
  rules={[
    {
      positive: {
        component: ComponentRuleWithLabelDo(),
        description: "Keep the label text in the menu item short and informative. Use 1 word only."
      },
      negative: {
        component: ComponentRuleWithLabelDont(),
        description: "Dont use long text with more than 1 word, to ensure clarity and to avoid ellipsis."
      }
    }
  ]}
/>
