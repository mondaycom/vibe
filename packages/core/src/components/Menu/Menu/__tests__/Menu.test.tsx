import { vi, beforeEach, afterEach, describe, it, expect } from "vitest";
import React from "react";
import renderer from "react-test-renderer";
import { cleanup, fireEvent, render } from "@testing-library/react";
import { act } from "@testing-library/react-hooks";
import Menu from "../Menu";
import MenuItem from "../../MenuItem/MenuItem";
import MenuTitle from "../../MenuTitle/MenuTitle";
import Divider from "../../../Divider/Divider";
import { mockRequestAnimationFrame, restoreRequestAnimationFrameMock } from "../../../../tests/__tests__/test-utils";

describe("Snapshots", () => {
  beforeEach(() => {
    mockRequestAnimationFrame();
  });

  afterEach(() => {
    restoreRequestAnimationFrameMock();
  });

  it("renders correctly with empty props", () => {
    const tree = renderer.create(<Menu />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders correctly with custom class name", () => {
    const tree = renderer.create(<Menu className="dummy-class-name" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders correctly with children", () => {
    const tree = renderer
      .create(
        <Menu className="dummy-class-name" focusItemIndexOnMount={0}>
          <MenuTitle caption="my title" />
          <MenuItem title="item 1" />
          <Divider />
          <MenuItem title="item 2" />
        </Menu>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

vi.useFakeTimers();
const menuTitleCaption = "Title";
const menuItem1Name = "My item 1";
const menuItem1Id = "menu-item-1";
const menuItem1OnClickMock = vi.fn();
const menuItem2OnClickMock = vi.fn();
const menuItem2Name = "My item 2";
const menuItem2Id = "menu-item-2";

const renderComponent = ({ ...props } = {}) => {
  return render(
    <Menu {...props} ariaLabel="menu">
      <MenuTitle caption={menuTitleCaption} />
      <MenuItem title={menuItem1Name} onClick={menuItem1OnClickMock} id={menuItem1Id} />
      <Divider />
      <MenuItem title={menuItem2Name} onClick={menuItem2OnClickMock} id={menuItem2Id} />
    </Menu>
  );
};

describe.skip("<Menu />", () => {
  afterEach(() => {
    cleanup();
  });

  it("calls onClick only for the selected menu item when using the mouse", () => {
    const menuComponent = renderComponent();

    const menuItem = menuComponent.getByText(menuItem1Name);

    act(() => {
      fireEvent.mouseOver(menuItem);
      vi.advanceTimersByTime(1000);
      fireEvent.click(menuItem);
    });

    vi.advanceTimersByTime(1000);
    expect(menuItem1OnClickMock.mock.calls.length).toBe(1);
    expect(menuItem2OnClickMock.mock.calls.length).toBe(0);
  });

  it("calls onClick only for the selected menu item when using the enter", () => {
    const menuComponent = renderComponent();
    const menuItem = menuComponent.getByText(menuItem1Name);

    act(() => {
      fireEvent.mouseOver(menuItem);
      vi.advanceTimersByTime(1000);
      fireEvent.keyUp(menuItem, { key: "Enter" });
    });

    vi.advanceTimersByTime(1000);
    expect(menuItem1OnClickMock.mock.calls.length).toBe(1);
    expect(menuItem2OnClickMock.mock.calls.length).toBe(0);
  });

  it("calls onClick only for the selected menu item when using keyboard", () => {
    const menuComponent = renderComponent();
    const menuElement = menuComponent.getByLabelText("menu");

    act(() => {
      fireEvent.keyUp(menuElement, { key: "ArrowDown" });
      vi.advanceTimersByTime(1000);
      fireEvent.keyUp(menuElement, { key: "Enter" });
    });

    vi.advanceTimersByTime(1000);
    expect(menuItem1OnClickMock.mock.calls.length).toBe(1);
    expect(menuItem2OnClickMock.mock.calls.length).toBe(0);
  });

  it("menu has correct active-descendant when item is active", () => {
    const menuComponent = renderComponent();
    expect(menuComponent).not.toHaveAttribute("aria-activedescendant");

    const menuItem = menuComponent.getByText(menuItem1Name);

    act(() => {
      fireEvent.mouseOver(menuItem);
      vi.advanceTimersByTime(1000);
      fireEvent.click(menuItem);
    });

    vi.advanceTimersByTime(1000);
    expect(menuComponent).toHaveAttribute("aria-activedescendant", menuItem1Id);
  });
});
