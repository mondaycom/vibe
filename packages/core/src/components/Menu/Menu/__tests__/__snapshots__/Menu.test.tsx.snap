// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshots renders correctly with children 1`] = `
<ul
  aria-label="Menu"
  className="menu medium dummy-class-name"
  data-testid="menu"
  id="menu-2"
  onBlur={[Function]}
  onFocus={[Function]}
  onMouseOver={[Function]}
  role="menu"
  tabIndex={0}
>
  <div
    className="typography secondary start singleLineEllipsis text text2Normal title"
    data-testid="menu-title_menu-2-item-0"
  >
    <label
      className="titleCaption titleCaptionBottom"
      data-testid="menu-title-caption_menu-2-item-0"
      id="menu-2-item-0"
    >
      my title
    </label>
  </div>
  <li
    aria-current={true}
    className="typography primary start singleLineEllipsis text text2Normal item focused initialSelected"
    data-testid="menu-item_1"
    id="menu-2-item-1"
    onClick={[Function]}
    role="menuitem"
    tabIndex={-1}
  >
    <div
      className="title"
    >
      item 1
    </div>
  </li>
  <li
    aria-current={false}
    className="typography primary start singleLineEllipsis text text2Normal item initialSelected"
    data-testid="menu-item_2"
    id="menu-2-item-2"
    onClick={[Function]}
    role="menuitem"
    tabIndex={-1}
  >
    <div
      className="title"
    >
      item 2
    </div>
  </li>
</ul>
`;

exports[`Snapshots renders correctly with custom class name 1`] = `
<ul
  aria-label="Menu"
  className="menu medium dummy-class-name"
  data-testid="menu"
  id="menu-1"
  onBlur={[Function]}
  onFocus={[Function]}
  onMouseOver={[Function]}
  role="menu"
  tabIndex={0}
/>
`;

exports[`Snapshots renders correctly with empty props 1`] = `
<ul
  aria-label="Menu"
  className="menu medium"
  data-testid="menu"
  id="menu-0"
  onBlur={[Function]}
  onFocus={[Function]}
  onMouseOver={[Function]}
  role="menu"
  tabIndex={0}
/>
`;
